crypto/keys/multisig/multisig.go

```diff
@@ -0,0 +1,236 @@
+	}
+	// index in the list of signatures which we are concerned with.
+	sigIndex := 0
+	for i := 0; i < size; i++ {
+		if bitarray.GetIndex(i) {
+			si := sig.Signatures[sigIndex]
+			switch si := si.(type) {
+			case *signing.SingleSignatureData:
+				msg, err := getSignBytes(si.SignMode)
+				if err != nil {
+					return err
+				}
+				if !pubKeys[i].VerifySignature(msg, si.Signature) {
+					return err
+				}
+			case *signing.MultiSignatureData:
+				nestedMultisigPk, ok := pubKeys[i].(multisigtypes.PubKey)
+				if !ok {
+					return fmt.Errorf("unable to parse pubkey of index %d", i)
+				}
+				if err := nestedMultisigPk.VerifyMultisignature(getSignBytes, si); err != nil {
+					return err
+				}
+			default:
+				return fmt.Errorf("improper signature data type for index %d", sigIndex)
+			}
+			sigIndex++
+		}
+	}
+	return nil
+}
+
+// VerifySignature implements crypto.PubKey VerifySignature method,
+// it panics because it can't handle MultiSignatureData
+// cf. https://github.com/cosmos/cosmos-sdk/issues/7109#issuecomment-686329936
+func (m *LegacyAminoPubKey) VerifySignature(msg []byte, sig []byte) bool {
+	panic("not implemented")
+}
+
+// GetPubKeys implements the PubKey.GetPubKeys method
+func (m *LegacyAminoPubKey) GetPubKeys() []tmcrypto.PubKey {
+	if m != nil {
+		pubKeys := make([]tmcrypto.PubKey, len(m.PubKeys))
+		for i := 0; i < len(m.PubKeys); i++ {
+			pubKeys[i] = m.PubKeys[i].GetCachedValue().(tmcrypto.PubKey)
+		}
+		return pubKeys
+	}
+
+	return nil
+}
+
+// Equals returns true if m and other both have the same number of keys, and
+// all constituent keys are the same, and in the same order.
+func (m *LegacyAminoPubKey) Equals(key tmcrypto.PubKey) bool {
+	otherKey, ok := key.(multisigtypes.PubKey)
+	if !ok {
+		return false
+	}
+	pubKeys := m.GetPubKeys()
+	otherPubKeys := otherKey.GetPubKeys()
+	if m.GetThreshold() != otherKey.GetThreshold() || len(pubKeys) != len(otherPubKeys) {
+		return false
+	}
+
+	for i := 0; i < len(pubKeys); i++ {
+		if !pubKeys[i].Equals(otherPubKeys[i]) {
+			return false
+		}
+	}
+	return true
+}
+
+// GetThreshold implements the PubKey.GetThreshold method
+func (m *LegacyAminoPubKey) GetThreshold() uint {
+	return uint(m.Threshold)
+}
+
+// Type returns multisig type
+func (m *LegacyAminoPubKey) Type() string {
+	return "PubKeyMultisigThreshold"
+}
+
+// UnpackInterfaces implements UnpackInterfacesMessage.UnpackInterfaces
+func (m *LegacyAminoPubKey) UnpackInterfaces(unpacker types.AnyUnpacker) error {
+	for _, any := range m.PubKeys {
+		var pk crypto.PubKey
+		err := unpacker.UnpackAny(any, &pk)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// pubKeyMultisigThreshold represents the old PubKeyMultisigThreshold. Now, we
+// convert the new one (LegacyAminoPubKey) to this one, and marshal this one.
+type pubKeyMultisigThreshold struct {
+	K       uint            `json:"threshold"`
+	PubKeys []crypto.PubKey `json:"pubkeys"`
+}
```

Comment:
I still kept the old struct, but as private in this file. Reason: for marshalling, we convert the proto `LegacyAminoPubKey` to this `pubKeyMultisigThreshold`, and use AminoCdc to marshal this old `pubKeyMultisigThreshold`. We do the reverse for unmarshalling.

I think that's the easiest way, right?

Reply:
Answering myself: no, the easiest is to just not add any Amino marshal overrides on the proto struct.

The only doubt I had: In the new one, K is a uint32, where it's a uint in the old one.